File: docker/ruby/srcs/server/app/channels/online_channel.rb
class OnlineChannel < ApplicationCable::Channel
  def subscribed
    stream_from "online_channel"
    if current_user
      ActionCable.server.broadcast "online_channel", { user: current_user.id, online: :on }
      current_user.online = true
      current_user.pong = 0
      current_user.save!
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
    if current_user
      ActionCable.server.broadcast "online_channel", { user: current_user.id, online: :off }
      current_user.online = false
      current_user.pong = 0
      current_user.save!
    end
  end
end

File: docker/ruby/srcs/server/app/javascript/channels/online_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("OnlineChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    if (data.online == "on" && Transcendence.users && Transcendence.users.get(data.user)) {
      Transcendence.users.get(data.user).set({online: true})
    } else if (data.online == "off" && Transcendence.users && Transcendence.users.get(data.user)) {
      Transcendence.users.get(data.user).set({online: false})
    }
  }
});
